### PR TITLE
Add feedback form and simple validation 

### PR DESCRIPTION
**Pull request by AndrewHelix**
Add form for feedback and simple validation for email and name. 
Feedback data save in localStorage
Screenshot 2022-09-15 182958

### CODE REVIEW

File: src/hooks/useValidate.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { useState } from 'react';
+
+const validationSchema: { [index: string]: (value: string) => boolean } = {
+  name: (value: string) => /[a-zA-Z0-9_]{3,}/gm.test(value),
+  email: (value: string) =>
+    /^(([^<>()[\].,;:\s@"]+(\.[^<>()[\].,;:\s@"]+)*)|(".+"))@(([^<>()[\].,;:\s@"]+\.)+[^<>()[\].,;:\s@"]{2,})$/iu.test(
```
Comment (JuliaShavyrova) [inline src/hooks/useValidate.ts]: please move regular expressions to separate variables

File: src/hooks/useValidate.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      value
+    )
+};
+
+export default function useValidate() {
+  const [isValid, setIsValid] = useState<boolean>(false);
+  const validateInput = (element: HTMLInputElement, value: string) => {
+    setIsValid(validationSchema[element.name](value));
```
Comment (JuliaShavyrova) [inline src/hooks/useValidate.ts]: no need to pass the whole element

File: src/components/feedback/Feedback.tsx [hunk 3]
```diff
@@ -1,11 +1,152 @@
...
+
+  const email = useValidate();
+  const name = useValidate();
+
+  const stateHandler = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
+    dispatch({
+      type: CHANGE_FEEDBACK_STATE,
+      payload: {
+        name: e.target.name,
+        data: e.target.type === 'range' ? +e.target.value : e.target.value
```
Comment (JuliaShavyrova) [inline src/components/feedback/Feedback.tsx]: can't you destructure `e.target`?

File: src/components/feedback/Feedback.tsx [hunk 4]
```diff
@@ -1,11 +1,152 @@
...
+    e.preventDefault();
+    setInLocalStorage(state, FEEDBACK_DATA_LS_LABEL);
+    resetForm();
+  };
+
+  const ratingColor = useMemo(() => {
+    if (state.rating >= 8) return 'green';
+    if (state.rating >= 5) return 'yellow';
+    if (state.rating < 5) return 'red';
+  }, [state.rating]);
```
Comment (JuliaShavyrova) [inline src/components/feedback/Feedback.tsx:85]: same question about destructuring state

File: src/components/feedback/Feedback.tsx [hunk 5]
```diff
@@ -1,11 +1,152 @@
...
 
 export const Feedback: FC = () => {
+  const [state, dispatch] = useReducer(feedbackReducer, {
+    name: '',
+    email: '',
+    problems: '',
+    rating: 5
+  });
+
+  const email = useValidate();
```
Comment (JuliaShavyrova) [inline src/components/feedback/Feedback.tsx]: please add types

File: src/components/feedback/Feedback.tsx [hunk 6]
```diff
@@ -1,11 +1,152 @@
...
+            type="submit"
+            className={`${styles['form-btn']} ${styles['send-btn']}`}
+            disabled={!email.isValid || !name.isValid}
+            onClick={sendForm}
+          >
+            Send
+          </button>
+          <button
+            type="reset"
+            className={`${styles['form-btn']} ${styles['reset-btn']}`}
```
Comment (JuliaShavyrova) [inline src/components/feedback/Feedback.tsx]: you can try to use classnames library

File: src/components/feedback/Feedback.tsx [hunk 7]
```diff
@@ -1,11 +1,152 @@
...
+    });
+  };
+
+  const requiredInputHandler = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.name === 'name') {
+      name.validateInput(e.target, e.target.value);
+    }
+    if (e.target.name === 'email') {
+      email.validateInput(e.target, e.target.value);
+    }
```
Comment (JuliaShavyrova) [inline src/components/feedback/Feedback.tsx:64]: conditions can be refactored. in any case you'll call `validateInput` method with the same params, using condition you can decide what control to select

File: src/components/feedback/Feedback.tsx [hunk 8]
```diff
@@ -1,11 +1,152 @@
...
+    dispatch({
+      type: CHANGE_FEEDBACK_STATE,
+      payload: {
+        name: e.target.name,
+        data: e.target.type === 'range' ? +e.target.value : e.target.value
+      }
+    });
+  };
+
+  const requiredInputHandler = (e: ChangeEvent<HTMLInputElement>) => {
```
Comment (JuliaShavyrova) [inline src/components/feedback/Feedback.tsx]: cannot say that it's a handler for 'required' check only, please rename

File: src/components/feedback/Feedback.tsx (hunk 7 above)
Comment (AndrewHelix) [reply src/components/feedback/Feedback.tsx:64]: but I need to call different `validateInput`. It depends on conditions

Comment (JuliaShavyrova) [reply src/components/feedback/Feedback.tsx:64]: as an option:
```
let selectedInput;
if (...) {
  selectedInput = ...
}
selectedInput?.validateInput(e.target, e.target.value);
```

File: src/components/feedback/Feedback.tsx [hunk 9]
```diff
@@ -1,43 +1,62 @@
...
+import { ValidateElementInfoType } from 'hooks/useValidate';
+import { getBindedStyles } from 'helpers';
+
+const getClasses = getBindedStyles(styles);
 
 export const Feedback: FC = () => {
-  const [state, dispatch] = useReducer(feedbackReducer, {
+  const [state, dispatch] = useReducer<
+    (state: FeedbackState, { type, payload }: FeedbackActionData) => FeedbackState
+  >(feedbackReducer, {
```
Comment (JuliaShavyrova) [inline src/components/feedback/Feedback.tsx]: you can use built-in type `Reducer` like here:
https://codesandbox.io/s/react-typescript-hooks-2-y7zvkl?file=/src/App/App.tsx:441-448

File: src/components/main/MainCard.tsx [hunk 10]
```diff
@@ -58,9 +60,14 @@ const MainCard: FC<MainCardProps> = ({ currentWeather, location }) => {
...
 
+  const mainCardClasses = useMemo(
+    () => getClasses('main-card__item', 'main-card__item_center'),
+    []
+  );
+
   return (
     <div className={styles['main-card']}>
-      <div className={`${styles['main-card__item']} ${styles['main-card__item_center']}`}>
+      <div className={mainCardClasses}>
```
Comment (JuliaShavyrova) [inline src/components/main/MainCard.tsx]: just interesting - doesn't classnames work without binding like this?
`<div className={classnames(styles['main-card__item'], styles['main-card__item_center'])}>`

**Review by JuliaShavyrova — APPROVED** [review]

File: src/components/main/MainCard.tsx (hunk 10 above)
Comment (AndrewHelix) [reply src/components/main/MainCard.tsx]: work;)
fix it?

Comment (JuliaShavyrova) [reply src/components/main/MainCard.tsx]: yes, no need in overengineering

**Pull request merged**